## Final Project Submission

Please fill out:
* Student name: Jessica Forrest-Baldini
* Student pace: Part-time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


## Import Packages & Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy.stats as stats
from datetime import datetime as date
import matplotlib.pyplot as plt
%matplotlib inline

data = pd.read_csv('kc_house_data.csv') 
df = pd.DataFrame(data)

## Clean Data

In [2]:
# Drop waterfront - only 146 homes with waterfronts
# Drop view - this represent if the property has been viewed
# Drop id

to_drop = ['waterfront','view','id']
df = df.drop(to_drop, axis=1)

In [3]:
df.isna().sum()

date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3842
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [4]:
# Since not all homes have been renovated, I'm going to replace the 'NaN' values for 
# 'yr_renovated' with '0'
df.yr_renovated = df.yr_renovated.fillna(0)

In [5]:
# Replace NaN basement values with 0 as there are only 454 of them 
# Test median values for these later to see if it improves model

df.sqft_basement = df.sqft_basement.replace('?','0.0').astype(float)
#df.sqft_basement = df.sqft_basement+1

In [6]:
# Take abs of longitude for normalization later on
df.long = abs(df.long)

In [7]:
df.describe().round()

price  bedrooms  bathrooms  sqft_living   sqft_lot   floors  \
count    21597.0   21597.0    21597.0      21597.0    21597.0  21597.0   
mean    540297.0       3.0        2.0       2080.0    15099.0      1.0   
std     367368.0       1.0        1.0        918.0    41413.0      1.0   
min      78000.0       1.0        0.0        370.0      520.0      1.0   
25%     322000.0       3.0        2.0       1430.0     5040.0      1.0   
50%     450000.0       3.0        2.0       1910.0     7618.0      2.0   
75%     645000.0       4.0        2.0       2550.0    10685.0      2.0   
max    7700000.0      33.0        8.0      13540.0  1651359.0      4.0   

       condition    grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
count    21597.0  21597.0     21597.0        21597.0   21597.0       21597.0   
mean         3.0      8.0      1789.0          286.0    1971.0          69.0   
std          1.0      1.0       828.0          440.0      29.0         364.0   
min          1.0      3.0       370.0            0.0    1900.0           0.0   
25%          3.0      7.0      1190.0            0.0    1951.0           0.0   
50%          3.0      7.0      1560.0            0.0    1975.0           0.0   
75%          4.0      8.0      2210.0          550.0    1997.0           0.0   
max          5.0     13.0      9410.0         4820.0    2015.0        2015.0   

       zipcode      lat     long  sqft_living15  sqft_lot15  
count  21597.0  21597.0  21597.0        21597.0     21597.0  
mean   98078.0     48.0    122.0         1987.0     12758.0  
std       54.0      0.0      0.0          685.0     27274.0  
min    98001.0     47.0    121.0          399.0       651.0  
25%    98033.0     47.0    122.0         1490.0      5100.0  
50%    98065.0     48.0    122.0         1840.0      7620.0  
75%    98118.0     48.0    122.0         2360.0     10083.0  
max    98199.0     48.0    123.0         6210.0    871200.0

We can see here that bedrooms has what seems to be a major outlier, 33 bedrooms. Let's take a deeper look. 

In [8]:
# Compare with other homes in data set that have high number of bedrooms
df[df.bedrooms > 9] 

date      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
8748    8/21/2014   520000.0        11       3.00         3000      4960   
13301   8/14/2014  1150000.0        10       5.25         4590     10920   
15147  10/29/2014   650000.0        10       2.00         3610     11914   
15856   6/25/2014   640000.0        33       1.75         1620      6000   
19239  12/29/2014   660000.0        10       3.00         2920      3745   

       floors  condition  grade  sqft_above  sqft_basement  yr_built  \
8748      2.0          3      7        2400          600.0      1918   
13301     1.0          3      9        2500         2090.0      2008   
15147     2.0          4      7        3010          600.0      1958   
15856     1.0          5      7        1040          580.0      1947   
19239     2.0          4      7        1860         1060.0      1913   

       yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
8748         1999.0    98106  47.5560  122.363           1420        4960  
13301           0.0    98004  47.5861  122.113           2730       10400  
15147           0.0    98006  47.5705  122.175           2040       11914  
15856           0.0    98103  47.6878  122.331           1330        4700  
19239           0.0    98105  47.6635  122.320           1810        3745

In [9]:
# 33 bedrooms and only 1.75 bathrooms doesn't seem right. It is a pretty big lot at 6,000sqft
# but 33 bedrooms doesn't seem right. Remove. 

to_drop = df[df.bedrooms == 33].index
df = df.drop(to_drop)

In [ ]:
df.describe().round()

We can see price has some big outliers as well, as the 75th percentile is \\$ 645,000 and they max is \\$ 7,700,000, which is \\$ 7M more. 

Let's take a closer look at the percentiles.

In [ ]:
for i in range(0,100):
    q = i/100
    print("{} percentile: {}".format(q, df.price.quantile(q=q)))

I'm going to remove the bottom and top percentile to remove outliers.

In [ ]:
orig_tot = len(df)
df = df[(df.price > 154000.0) & (df.price < 1970000.0)] # Subsetting to remove extreme outliers
print('Percent removed:', (orig_tot -len(df))/orig_tot)

In [ ]:
df.head()

In [ ]:
for i in range(0,100):
    q = i/100
    print("{} percentile: {}".format(q, df.sqft_lot.quantile(q=q)))

In [ ]:
orig_tot = len(df)
df = df[(df.sqft_lot > 1010.61) & (df.sqft_lot < 213008.0)] # Subsetting to remove extreme outliers
print('Percent removed:', (orig_tot -len(df))/orig_tot)

In [ ]:
#print(df.sqft_basement.describe())
#df.sqft_basement.value_counts()
#df[df.sqft_basement > 100].sort_values(by='sqft_basement', ascending=True).head(20)
print(sum(df.sqft_basement > 0))
#sns.regplot(df.sqft_basement,df.price);
sns.regplot(df[df.sqft_basement >= 110].sqft_basement,df[df.sqft_basement >= 110].price);

Only 8,317 of the 32,000+ homes have basements, so we're going to bin them. I used zero as one bin and then did a quartile split for all the 'sqft_basement' values for homes with basements. 

In [ ]:
# # Bin 'sqft_basement' starting at the min square footage for homes with a basement
# bins = pd.qcut(df.sqft_basement[df.sqft_basement >= 110],q=4)
# bins.value_counts()

In [ ]:
# Bin sqft_basement

bins_sqft_basement = [0,109,450,980,5000]

# Bin data & return dummies
def binned_dummies(data, features, bins):
    data_bins = pd.cut(data, bins)
    data_bins = data_bins.cat.as_unordered()
    dummies = pd.get_dummies(data_bins, prefix = features, drop_first=True)
    return dummies

dummies_sqft_basement = binned_dummies(df.sqft_basement,'sqft_basement', bins_sqft_basement)

# Remove original column from data set
df = df.drop(['sqft_basement'], axis=1)
                                            
# Add new columns in
df = pd.concat([df, dummies_sqft_basement], axis=1)

In [ ]:
# plt.hist(df.grade);

In [ ]:
# plt.hist(df.condition);

In [ ]:
# df.condition.describe()

In [ ]:
# plt.hist(df.long);

In [ ]:
# Define function to convert datestr to datenum
def datenum(datestr):
    '''
    Convert datestring in the format MM/DD/YYYY
    to MATLAB style datenum
    '''
    datenum = date.toordinal(date((int(datestr.split('/', -1)[2])),
                                  (int(datestr.split('/', -1)[0])),
                                  (int(datestr.split('/', -1)[1]))
                                 ))+366
    return datenum

# Apply to date column
df.date = df.date.map(datenum)

Explore the data and check for any outliers.

In [ ]:
# round(df.describe(),1)

'sqft_lot' seems to have a major outlier at 1,651,359 sqft. Let's take a closer look.

In [ ]:
# print(df.sqft_lot.median())
# display(df.sort_values(by='sqft_lot', ascending=False).head(5))

This appears to be the highest, but not necessarily an outlier. Let's go ahead and leave it for now. 

In [ ]:
# df.long.describe()

In [ ]:
# plt.hist(df.long);

## Normalize, MinMax Scale, Standardize, 
One-Hot Encode

First, check for normality, heteroscedasticity & discover categorical data. Can get a good idea of categorical features from looking at the data above, but let's explore normality, categorical and any relationships using a pairplot.

In [ ]:
# Commented out because takes a long time to run
#sns.pairplot(df)

While 'condition' and 'grade' are technically categorical, they are on a scale, so I am going to leave them as is, and will min-max scale them. 

In [ ]:
# Normalize Data

# Log Transform 

# Continuous variables
features = ['date','price','sqft_living','lat','long','bedrooms','bathrooms','floors','condition','grade',
            'sqft_lot','sqft_above','yr_built',
            'sqft_living15','sqft_lot15','zipcode']

df_cont_features = df[features]

# Add '_log' to continuous variable column names
log_names = [f'{column}_log' for column in df_cont_features.columns]

# Log transform continuous variables
df_log = np.log(df_cont_features)
df_log.columns = log_names


### Normalize (subract mean and divide by std)

# Define function to normalize
def normalize(feature):
    return (feature - feature.mean()) / feature.std()

# Apply function to normalize
df_log_norm = df_log.apply(normalize)

# # Define function to min-max scale
# def minmaxscale(feature):
#     return (feature-min(feature))/(max(feature)-min(feature))

# # Apply function to min-max scale
# df_log_norm_scale = df_log.apply(minmaxscale)

# Remove original column from data set
df = df.drop(features, axis=1)

# Add new columns in
df = pd.concat([df, df_log_norm], axis=1) #_scale

In [ ]:
# df.head()

## Check for Multicollinearity 

### Multicollinearity

In [ ]:
# Take a look at the correlation matrix to check for multicollinearity
abs(df.corr()) > 0.75

In [ ]:
# # Check & see if sqft_lot15 & sqft_living15 are correlated with sqft_lot & sqft_living
# sns.regplot(df.sqft_lot,df.sqft_lot15)
# plt.show()
# sns.regplot(df.sqft_living,df.sqft_living15)
# plt.show()

We can see that 'sqft_lot15' & 'sqft_living15' are correlated with 'sqft_lot' & 'sqft_living', which makes sense because the features appended with 15 represent the average of the 15 nearest neighbors. 

We will remove these values to prevent multicollinearity in our model. 

In [ ]:
# # Remove sqft_lot15 & sqft_living15
# df = df.drop(['sqft_lot15','sqft_living15'], axis=1)

In [ ]:
# # Take a look at the correlation matrix to check for multicollinearity
# df.corr() > 0.75

We can see here that 'sqft_living' seems to be highly correlated with multiple features such as 'bathrooms', 'grade', and 'sqft_above'. 'sqft_living' represents the square footage of the entire home and seems that it would be a strong indicator of home price. The squarefoot of other features are essentially subsets of the overall home squarefootage. So for now I am going to remove it. However, it could be kept in place and other features removed later on to see if this improves the overall performance of the model.

In [ ]:
# # Remove 'sqft_living' to prevent multicollinearity 
# # as it is highly correlated with multiple features

# df = df.drop(['sqft_living'], axis=1)

In [ ]:
# abs(df.corr()) > 0.75

It looks like 'sqft_above' and 'grade' are highly correlated, so we should only keep one to prevent multicollinearity. 

For now I'm going to keep 'sqft_above', but can also test removing 'sqft_above' and one-hot-encoding 'grade' later on to see if it improves performance of the model. 

In [ ]:
# # We can see the relationship here between 'sqft_above' and 'grade', it appears to possibly
# # be polynomial. While grade is categorical, it is continuous
# sns.regplot(df.grade,df.sqft_above);

Since the relationship looks slightly polynomial, and 'grade' technically is categorical, I want to look at their distributions to further explore the relationship. Let's look at the histograms/distplot.

In [ ]:
# sns.distplot(df.sqft_above)
# plt.show()
# plt.hist(df.grade)
# plt.show()

Both do appear to be skewed in the same direction, so perhaps they are linearly related. Okay, I will remove grade. 

In [ ]:
# df = df.drop(['grade'], axis=1)

In [ ]:
# df.head()

## Start Modeling

### Split train/test datasets

In [ ]:
# Separate target and feature variables

X = df.drop(['price_log'],axis=1)
y = df.price_log

In [ ]:
# Import train_test_split from sklearn
from sklearn.model_selection import train_test_split

# Split data with test size of 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

### Fit Model

In [ ]:
from sklearn.linear_model import LinearRegression

# Initialize the linear regression model class
linreg = LinearRegression()

# Fit the model to train data
linreg.fit(X_train, y_train)

# Calculate predictions on test set
y_hat_test = linreg.predict(X_test)

### Cross Validate Model

In [ ]:
from sklearn.metrics import mean_squared_error, make_scorer
from sklearn.model_selection import cross_val_score

mse = make_scorer(mean_squared_error)

# Test Errors Results
cv_5_results = cross_val_score(linreg, X, y, cv=5, scoring=mse)
cv_5_results


print(f"RMSE: {np.sqrt(cv_5_results.mean())}")

In [ ]:
from sklearn.model_selection import KFold

crossval = KFold(n_splits=3, shuffle=True, random_state=1)

baseline_R2 = np.mean(cross_val_score(linreg, X, y, scoring='r2', cv=crossval))
baseline_R2

### Statsmodel (OLS)

In [ ]:
import statsmodels.api as sm
X = sm.add_constant(X)
model = sm.OLS(y,X)
results = model.fit()

results.summary()

### Interactions

In [ ]:
from itertools import combinations

# Find top interactions by R^2 value

# Use combinations from itertools to create all possible combinations of two features
feat_combinations = combinations(X_train.columns, 2)

# Empty list to fill for interactons values
interactions = []

# for i, (feature1,feature2) in feature_combinations:
for i, (a, b) in enumerate(feat_combinations):
    # fill interatctions list with feature a * feature b
    X_train['interaction'] = X_train[a] * X_train[b]
    R2 = np.mean(cross_val_score(linreg, X_train, y_train, scoring='r2', cv=crossval))
    if R2 > baseline_R2:
        interactions.append((a, b, round(R2,5)))
            
print('Top 3 interactions: %s' %sorted(interactions, key=lambda inter: inter[2], reverse=True)[:3])

# Best Performing Model - So Far

In [ ]:
#Build a final model with interactions
#Use 10-fold cross-validation to build a model using the above interaction.

crossval = KFold(n_splits=10, shuffle=True, random_state=1)
final = X.copy()

final['lat_log*sqft_lot15_log'] = final['lat_log'] * final['sqft_lot15_log']
#final['lat_log*sqft_lot_log'] = final['lat_log'] * final['sqft_lot_log']
final['long_log*zipcode_log'] = final['long_log'] * final['zipcode_log']

final_model_R2 = np.mean(cross_val_score(linreg, final, y, scoring='r2', cv=crossval))

print(final_model_R2)

import statsmodels.api as sm
df_inter_sm = sm.add_constant(final)
model = sm.OLS(y,final)
results = model.fit()

results.summary()

In [ ]:
# Test Errors Results
cv_5_results = cross_val_score(linreg, final, y, cv=5, scoring=mse)
cv_5_results


print(f"RMSE: {np.sqrt(cv_5_results.mean())}")

In [ ]:
fig = sm.graphics.qqplot(results.resid, dist=stats.norm, line='45', fit=True)